addons/udes_stock/wizard/udes_proactive_replens.py

```diff
@@ -0,0 +1,88 @@
+from odoo import models, fields, _
+from odoo.exceptions import ValidationError
+
+
+class UdesProactiveReplen(models.TransientModel):
+    _name = "udes.proactive.replen"
+    _description = "UDES Proactive Replenishment"
+
+    number_replen_moves = fields.Integer(
+        "Number of Products",
+        default=20,
+        help=("Number of products to generate proactive replen moves."
+        ),
+    )
+    replen_percentage = fields.Integer(
+        "Replen Percentage above the minimum",
+        default=20,
+        help=("Percentage quantity above the minimum to trigger proactive replen."
+        ),
+    )
+
+    def generate_replen_moves(self):
+        """Generate replen moves
+        """
+        OrderPoint = self.env["stock.warehouse.orderpoint"]
+        Picking = self.env["stock.picking"]
+        ProcurementGroup = self.env["procurement.group"]
+        Quant = self.env["stock.quant"]
+        pickings = Picking.browse()
+        picking_type = self.env.ref("udes_stock.picking_type_proactive_replen")
+
+        # Get all order points
+        domain = ProcurementGroup._get_orderpoint_domain()
+        orderpoints = OrderPoint.with_context(prefetch_fields=False).search(domain)
+        op_todo = {}
+        for orderpoint in orderpoints:
+            if orderpoint.qty_to_order > 0:
+                # This will need a normal replen, so would skip
+                continue
+            top = orderpoint.product_max_qty - orderpoint.product_min_qty
+            if top <= 0:
+                continue
+            # orderpoint.qty_forecast takes into account both normal replens and proactive replens,
+            # therefore "extra" will include them if they exist and a proactive replen won't get
+            # created
+            extra = orderpoint.qty_forecast - orderpoint.product_min_qty
+            percent_above_min = extra / top * 100
+            if percent_above_min <= self.replen_percentage:
+                op_todo[orderpoint] = percent_above_min
+
+        counter = 0
+        for orderpoint, percent_above_min in sorted(op_todo.items(), key=lambda x: x[1], reverse=False):
+            if counter >= self.number_replen_moves:
+                break
+            quants = Quant.search(
```

Comment:
Is it worth hoisting this query out of the loop, fetching quants for all the products at once and filtering inside the loop?

Reply:
We could potentially have lots of order points, so I was trying to avoid to get all the quants for 1000 products, and only get the 20 or 30 we really need.